src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,460 @@
+
+    // Push signals and controls for future batch read/write
+    void GPUActivityAgent::init(int level, const std::vector<int> &fan_in, bool is_level_root)
+    {
+        m_gpu_frequency_requests = 0;
+        m_f_max = 0;
+        m_f_efficient = 0;
+        m_f_range = 0;
+
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            m_gpu_active_region_start.push_back(0.0);
+            m_gpu_active_region_stop.push_back(0.0);
+            m_gpu_active_energy_start.push_back(0.0);
+            m_gpu_active_energy_stop.push_back(0.0);
+        }
+
+        if (level == 0) {
+            init_platform_io();
+        }
+    }
+
+    void GPUActivityAgent::init_platform_io(void)
+    {
+
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            m_gpu_freq_status.push_back({m_platform_io.push_signal("GPU_FREQUENCY_STATUS",
+                                         GEOPM_DOMAIN_BOARD_ACCELERATOR,
+                                         domain_idx), NAN});
+            m_gpu_compute_activity.push_back({m_platform_io.push_signal("GPU_COMPUTE_ACTIVITY",
+                                              GEOPM_DOMAIN_BOARD_ACCELERATOR,
+                                              domain_idx), NAN});
+            m_gpu_utilization.push_back({m_platform_io.push_signal("GPU_UTILIZATION",
+                                         GEOPM_DOMAIN_BOARD_ACCELERATOR,
+                                         domain_idx), NAN});
+            m_gpu_energy.push_back({m_platform_io.push_signal("GPU_ENERGY",
+                                    GEOPM_DOMAIN_BOARD_ACCELERATOR,
+                                    domain_idx), NAN});
+        }
+
+        m_time = {m_platform_io.push_signal("TIME", GEOPM_DOMAIN_BOARD, 0), NAN};
+
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            m_gpu_freq_control.push_back(control{m_platform_io.push_control("GPU_FREQUENCY_CONTROL",
+                                         GEOPM_DOMAIN_BOARD_ACCELERATOR,
+                                         domain_idx), NAN});
+        }
+
+        auto all_names = m_platform_io.control_names();
+        if (all_names.find("DCGM::FIELD_UPDATE_RATE") != all_names.end()) {
+            // DCGM documentation indicates that users should query no faster than 100ms
+            // even though the interface allows for setting the polling rate in the us range.
+            // In practice reducing below the 100ms value has proven functional, but should only
+            // be attempted if there is a proven need to catch short phase behavior that cannot
+            // be accomplished with the default settings.
+            m_platform_io.write_control("DCGM::FIELD_UPDATE_RATE", GEOPM_DOMAIN_BOARD, 0, 0.1); //100ms
+            m_platform_io.write_control("DCGM::MAX_STORAGE_TIME", GEOPM_DOMAIN_BOARD, 0, 1);
+            m_platform_io.write_control("DCGM::MAX_SAMPLES", GEOPM_DOMAIN_BOARD, 0, 100);
+        }
```

Comment:
Thinking about this code block when we're using LevelZero and not NVML...

I'm wondering if this code that effectively initializes the specific IOGroup (DCGM in this case) actually belongs inside the IOGroup itself.  It struck me as odd that the agent here needs to understand the specifics of the IOGroup that is currently compiled in.

Reply:
For this agent the current default values used by the DCGMDevicePool in dev should suffice, but the user may want to change them if their workload has shorter phases.

We could remove this code if we both assume that a user will do a `geompwrite DCGM::FIELD_UPDATE_RATE board 0 <better value>` prior to their if there's a known better value _and the DCGM setting is persistent for the duration of their time on the system_.  Whenever the dcgmWatchFields function is called in the DCGMDevicePool it needs to provide the three values being written, and prior to the GEOPM service I don't believe there a way to modify those outside the agent that would persist until the agent ran.